### PR TITLE
[GEOT-6444]: Add units to DistanceBufferOperator

### PR DESCRIPTION
**Pull request by mglazer**
This adds in a new type, `DistanceUnits` which is used for passing around the distance and units, as specified in the specification.

### CODE REVIEW

File: modules/extension/xsd/xsd-filter/src/main/java/org/geotools/filter/v1_0/OGCDistanceBufferTypeBinding.java [hunk 1]
```diff
@@ -74,41 +74,6 @@ public Class getType() {
...
     public void initialize(ElementInstance instance, Node node, MutablePicoContainer context) {}
 
-    /**
-     *
-     * <!-- begin-user-doc -->
-     * <!-- end-user-doc -->
-     *
-     * @generated modifiable
-     */
-    public Object parse(ElementInstance instance, Node node, Object value) throws Exception {
```
Comment (mglazer) [inline modules/extension/xsd/xsd-filter/src/main/java/org/geotools/filter/v1_0/OGCDistanceBufferTypeBinding.java]: This was unused so I just deleted it.

File: modules/library/opengis/src/main/java/org/opengis/filter/spatial/DistanceBufferOperator.java [hunk 2]
```diff
@@ -18,7 +18,9 @@
  * buffer around another shape. This could be used to find, for example, all the geometries that
  * come within 10 meters of a river.
  *
- * @version <A HREF="http://www.opengis.org/docs/02-059.pdf">Implementation specification 1.0</A>
+ * @version <A
```
Comment (mglazer) [inline modules/library/opengis/src/main/java/org/opengis/filter/spatial/DistanceBufferOperator.java]: Fixing a broken link

**Comment by ianturton** [comment]
Is there any way to leverage our existing units library (https://docs.geotools.org/stable/userguide/library/opengis/unit.html) to make sure that the units element maps to a real unit?

**Comment by mglazer** [comment]
So I'm having trouble parsing that web page, since it doesn't seem to map to what's actually in the source code.

The source only has a single constructor for measure:

```java
    public Measure(final double value, final Unit<?> unit) {
        this.value = value;
        this.unit = unit;
    }
```

but the docs say:

```java
Quantity<Length> afternoonWalk = Measure.valueOf( 6, NonSI.MILE );
```

Is there another class here (and Measure is not of type Quanity). I suspect the problem is that the ``unit.rst`` docs weren't upgraded (happy to submit a fix here):

https://docs.geotools.org/stable/userguide/welcome/upgrade.html#geotools-20-x

The problem though is that the units piece really seems to just be a string, at least according to the specification:

https://cite.opengeospatial.org/teamengine/about/wfs/1.0.0/site/OGCTestData/wfs/1.0.0/specs/filter/1.0.0/02-059.html#N10804

So I can try mapping it to a physical unit, but it seems implementation dependent on what the responding server actually expects you to give it (so I couldn't just give it the SI representation of 'm' if it expects a URI). 

Happy to try it out though, but just wanted to express my hesitation first. The units class definitely looks cool though, I never knew that existed for Java.

**Comment by ianturton** [comment]
OK So that was a longer diversion than I expected (good thing it's the holidays here) - so it seems that the Length units that we could consider here (from a units perspective) are 
And you are right the units page needs an update.

```
amstrong 1.0E10 amstrong
yd 1.0936132983377078 yd
mi 6.213711922373339E-4 mi
m 1.0 m
ft 3.2808398950131235 ft
ft_survey_us 3.2808333333333333 ft_survey_us
parsec 3.240779900164534E-17 parsec
in 39.37007874015748 in
UA 6.684587108863334E-12 UA
ly 1.0570234105227022E-16 ly
nmi 5.399568034557236E-4 nmi
```

Based on this code:

```
import java.util.HashSet;
import java.util.Set;

import javax.measure.Unit;
import javax.measure.quantity.Length;

import si.uom.NonSI;
import si.uom.SI;
import systems.uom.common.USCustomary;
import tec.uom.se.format.SimpleUnitFormat;
import tec.uom.se.format.SimpleUnitFormat.Flavor;
import tec.uom.se.quantity.Quantities;

public class LookupDistanceUnits {

  public static void main(String[] args) {
    javax.measure.Quantity<Length> a = Quantities.getQuantity(1.0, SI.METRE);
    Set<Unit<?>> units = new HashSet<>();
    units.addAll(NonSI.getInstance().getUnits());
    units.addAll(SI.getInstance().getUnits());
    units.addAll(USCustomary.getInstance().getUnits());
    SimpleUnitFormat formatter = SimpleUnitFormat.getInstance(Flavor.Default);
    formatter.label(NonSI.ANGSTROM, "amstrong");
    formatter.label(NonSI.PARSEC, "parsec");
    for (Unit<?> u : units) {
      if (u.isCompatible(SI.METRE)) {
        System.out.println(formatter.format(u) + " " + a.to((Unit<Length>) u));
      }
    }
  }
}
```

So realistically you could look for m, yd, mi, ft and nmi - and probably kilometres (handled as as a multiple of metre)

**Comment by mglazer** [comment]
OK, updated with that fix. For encoding, I ended up deciding to be opinionated and just use SI metres as the units, always, because units like "mi" didn't seem to have a `getSymbol()` which returned anything but `null`. Minimally I validate that the passed in unit is a length so this should be fine (and support almost everything under the sun).

**Comment by mglazer** [comment]
Docs update: https://github.com/geotools/geotools/pull/2729

**Comment by aaime** [comment]
This change is breaking the build in two GeoSever modules, wps and cross-layer-filtering.

Here is an example of a failure:

```
java.lang.RuntimeException: Parsing failed for Distance: javax.measure.format.ParserException: Parse Error
	at org.geotools.xsd.impl.ParseExecutor.visit(ParseExecutor.java:153)
	at org.geotools.xsd.impl.BindingWalker$BindingExecutionChain.execute(BindingWalker.java:220)
	at org.geotools.xsd.impl.BindingWalker.walk(BindingWalker.java:186)
	at org.geotools.xsd.impl.ElementHandlerImpl.endElement(ElementHandlerImpl.java:232)
	at org.geotools.xsd.impl.ParserHandler.endElement(ParserHandler.java:778)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.endElement(AbstractSAXParser.java:609)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEndElement(XMLDocumentFragmentScannerImpl.java:1782)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2967)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:602)
	at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:112)
	at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:505)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:842)
	at com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:771)
	at com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1213)
	at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:643)
	at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl.parse(SAXParserImpl.java:327)
	at org.geotools.xsd.Parser.parse(Parser.java:201)
	at org.geotools.xsd.Parser.parse(Parser.java:149)
	at org.geoserver.wfs.xml.v1_0_0.WfsXmlReader.read(WfsXmlReader.java:75)
	at org.geoserver.ows.Dispatcher.parseRequestXML(Dispatcher.java:1600)
	at org.geoserver.ows.Dispatcher.dispatch(Dispatcher.java:701)
	at org.geoserver.ows.Dispatcher.handleRequestInternal(Dispatcher.java:255)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:52)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1038)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:942)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:998)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:901)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:688)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:875)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:770)
	at org.geoserver.test.GeoServerSystemTestSupport$2.service(GeoServerSystemTestSupport.java:1654)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:770)
	at org.springframework.mock.web.MockFilterChain$ServletFilterProxy.doFilter(MockFilterChain.java:166)
	at org.springframework.mock.web.MockFilterChain.doFilter(MockFilterChain.java:133)
	at org.geoserver.test.GeoServerSystemTestSupport.dispatch(GeoServerSystemTestSupport.java:1683)
	at org.geoserver.test.GeoServerSystemTestSupport.dispatch(GeoServerSystemTestSupport.java:1594)
	at org.geoserver.test.GeoServerSystemTestSupport.dispatch(GeoServerSystemTestSupport.java:1582)
	at org.geoserver.test.GeoServerSystemTestSupport.postAsServletResponse(GeoServerSystemTestSupport.java:1230)
	at org.geoserver.test.GeoServerSystemTestSupport.postAsServletResponse(GeoServerSystemTestSupport.java:1175)
	at org.geoserver.test.GeoServerSystemTestSupport.post(GeoServerSystemTestSupport.java:1160)
	at org.geoserver.test.GeoServerSystemTestSupport.postAsDOM(GeoServerSystemTestSupport.java:1508)
	at org.geoserver.test.GeoServerSystemTestSupport.postAsDOM(GeoServerSystemTestSupport.java:1479)
	at org.geoserver.filter.function.WFSFilteringTest._testSingleSmallDistance(WFSFilteringTest.java:107)
	at org.geoserver.filter.function.WFSFilteringTest.testSingleSmallDistance10(WFSFilteringTest.java:96)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: javax.measure.format.ParserException: Parse Error
	at tec.uom.se.format.SimpleUnitFormat$DefaultFormat.check(SimpleUnitFormat.java:576)
	at tec.uom.se.format.SimpleUnitFormat$DefaultFormat.parseSingleUnit(SimpleUnitFormat.java:429)
	at tec.uom.se.format.SimpleUnitFormat$DefaultFormat.parseProductUnit(SimpleUnitFormat.java:440)
	at tec.uom.se.format.SimpleUnitFormat.parseObject(SimpleUnitFormat.java:269)
	at tec.uom.se.format.SimpleUnitFormat$DefaultFormat.parse(SimpleUnitFormat.java:782)
	at tec.uom.se.format.SimpleUnitFormat$DefaultFormat.parse(SimpleUnitFormat.java:777)
	at tec.uom.se.format.SimpleUnitFormat$DefaultFormat.parse(SimpleUnitFormat.java:768)
	at org.geotools.measure.Units.parseUnit(Units.java:238)
	at org.geotools.filter.v1_0.OGCDistanceTypeBinding.parse(OGCDistanceTypeBinding.java:96)
	at org.geotools.xsd.impl.ParseExecutor.visit(ParseExecutor.java:144)
	... 72 more
```

The distance filter in question has:

```<ogc:Distance units="meter"```

I believe the parsing got too strict now, here is an example from the Filter 1.1 specification (checked 1.0 and 2.0 too, the wording is the same, the example may vary):

> The units attribute is of type anyURI so that it may
> be used to reference a units dictionary. For example, the following XML fragment:
> <Distance unit=”http://www.uomdict.com/uom.html#meters”>10</Distance>
> encodes a distance value of 10 meters.

Reverting this change, needs to go back to the drawing board and accept more than just the units parseable by the units library.

**Comment by aaime** [comment]
The revert brought the build back to "blue": https://build.geoserver.org/job/geoserver-master/

Given the situation with the standard, with no indication of the actual list of possible values, I'm guessing keep on using a string instead of forcing a parseable unit is probably the better way to go.

**Comment by mglazer** [comment]
Well, that answers that question. Sorry about the break @aaime. I've pushed up another commit [here](https://github.com/geotools/geotools/pull/2734) which goes to back to a state where we just allow for random strings (as the spec states, but I think we were hoping for otherwise).

**Pull request merged**